### PR TITLE
WWST-4444 merge the wistar DTH to zigbee-window-shade

### PR DESCRIPTION
**Pull request by BYDengS**
@dkirker and @greens , please help review this commit.

### CODE REVIEW

File: devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy
Comment (dkirker) [inline devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy]: The only thing that maybe should change is to do `refresh() + cmds` so that we can get initial states as quickly as possible.

**Review by dkirker — DISMISSED** [review]

**Review by dkirker — DISMISSED** [review]

**Comment by dkirker** [comment]
@BYDengS For the sake of timeliness, I made a few commits.

**Review by dkirker — DISMISSED** [review]

File: devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy]: Rather than checking for a particular model which likely won't scale well, you could add functions like `usesLevelControl` and/or `usesWindowCovering`. Then when more manufacturers are added you just add update that one function rather than having to update multiple parts of the code.

File: devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy]: The level control cluster isn't listed in the fingerprint. Does it really support it?

File: devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy:126]: Does this work for wistar?

**Comment by dkirker** [comment]
@BYDengS Have a look at Tom's questions above then I'll go ahead and merge this.

File: devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy
Comment (BYDengS) [reply devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy:126]: yes,it works

File: devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy
Comment (BYDengS) [reply devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy]: yes, please see the log:
`
 dfd81aaa-030d-48e1-8be9-3a66d10d198e 上午9:45:42: debug description:- read attr - raw: BBB90E000808000020FF, dni: BBB9, endpoint: 0E, cluster: 0008, size: 8, attrId: 0000, result: success, encoding: 20, value: ff
dfd81aaa-030d-48e1-8be9-3a66d10d198e 上午9:45:43: debug description:- on/off: 1
dfd81aaa-030d-48e1-8be9-3a66d10d198e 上午9:45:42: debug description:- read attr - raw: BBB90E000808000020F2, dni: BBB9, endpoint: 0E, cluster: 0008, size: 8, attrId: 0000, result: success, encoding: 20, value: f2
dfd81aaa-030d-48e1-8be9-3a66d10d198e 上午9:45:42: debug description:- read attr - raw: BBB90E000808000020CC, dni: BBB9, endpoint: 0E, cluster: 0008, size: 8, attrId: 0000, result: success, encoding: 20, value: cc
dfd81aaa-030d-48e1-8be9-3a66d10d198e 上午9:45:41: debug description:- read attr - raw: BBB90E000808000020A8, dni: BBB9, endpoint: 0E, cluster: 0008, size: 8, attrId: 0000, result: success, encoding: 20, value: a8
dfd81aaa-030d-48e1-8be9-3a66d10d198e 上午9:45:41: debug description:- read attr - raw: BBB90E00080800002084, dni: BBB9, endpoint: 0E, cluster: 0008, size: 8, attrId: 0000, result: success, encoding: 20, value: 84
dfd81aaa-030d-48e1-8be9-3a66d10d198e 上午9:45:40: debug description:- read attr - raw: BBB90E0008080000205E, dni: BBB9, endpoint: 0E, cluster: 0008, size: 8, attrId: 0000, result: success, encoding: 20, value: 5e
dfd81aaa-030d-48e1-8be9-3a66d10d198e 上午9:45:39: debug description:- read attr - raw: BBB90E00080800002014, dni: BBB9, endpoint: 0E, cluster: 0008, size: 8, attrId: 0000, result: success, encoding: 20, value: 14
dfd81aaa-030d-48e1-8be9-3a66d10d198e 上午9:45:39: debug description:- catchall: 0104 0102 0E 01 0000 00 BBB9 00 00 0000 0B 01 0000
dfd81aaa-030d-48e1-8be9-3a66d10d198e 上午9:45:39: debug description:- on/off: 1
dfd81aaa-030d-48e1-8be9-3a66d10d198e 上午9:45:38: info open()
dfd81aaa-030d-48e1-8be9-3a66d10d198e 上午9:45:30: debug description:- read attr - raw: BBB90E00080800002000, dni: BBB9, endpoint: 0E, cluster: 0008, size: 8, attrId: 0000, result: success, encoding: 20, value: 00
dfd81aaa-030d-48e1-8be9-3a66d10d198e 上午9:45:30: debug description:- on/off: 1
dfd81aaa-030d-48e1-8be9-3a66d10d198e 上午9:45:30: debug description:- read attr - raw: BBB90E00080800002014, dni: BBB9, endpoint: 0E, cluster: 0008, size: 8, attrId: 0000, result: success, encoding: 20, value: 14
dfd81aaa-030d-48e1-8be9-3a66d10d198e 上午9:45:29: debug description:- read attr - raw: BBB90E00080800002038, dni: BBB9, endpoint: 0E, cluster: 0008, size: 8, attrId: 0000, result: success, encoding: 20, value: 38
dfd81aaa-030d-48e1-8be9-3a66d10d198e 上午9:45:29: debug description:- read attr - raw: BBB90E0008080000205E, dni: BBB9, endpoint: 0E, cluster: 0008, size: 8, attrId: 0000, result: success, encoding: 20, value: 5e
dfd81aaa-030d-48e1-8be9-3a66d10d198e 上午9:45:28: debug description:- read attr - raw: BBB90E00080800002084, dni: BBB9, endpoint: 0E, cluster: 0008, size: 8, attrId: 0000, result: success, encoding: 20, value: 84
dfd81aaa-030d-48e1-8be9-3a66d10d198e 上午9:45:28: debug description:- read attr - raw: BBB90E000808000020A8, dni: BBB9, endpoint: 0E, cluster: 0008, size: 8, attrId: 0000, result: success, encoding: 20, value: a8
dfd81aaa-030d-48e1-8be9-3a66d10d198e 上午9:45:27: debug description:- catchall: 0104 0102 0E 01 0000 00 BBB9 00 00 0000 0B 01 0100
dfd81aaa-030d-48e1-8be9-3a66d10d198e 上午9:45:27: debug description:- on/off: 0
dfd81aaa-030d-48e1-8be9-3a66d10d198e 上午9:45:27: info close()
dfd81aaa-030d-48e1-8be9-3a66d10d198e 上午9:45:00: debug description:- catchall: 0000 8021 00 00 0000 00 BBB9 00 00 0000 00 00 6000
dfd81aaa-030d-48e1-8be9-3a66d10d198e 上午9:44:58: debug description:- read attr - raw: BBB90E00080A00000020BC, dni: BBB9, endpoint: 0E, cluster: 0008, size: 10, attrId: 0000, result: success, encoding: 20, value: bc
dfd81aaa-030d-48e1-8be9-3a66d10d198e 上午9:44:58: info refresh()
dfd81aaa-030d-48e1-8be9-3a66d10d198e 上午9:44:58: debug Configuring Reporting and Bindings.
dfd81aaa-030d-48e1-8be9-3a66d10d198e 上午9:44:58: info configure() 
`

Comment (BYDengS) [reply devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy]: @tpmanley  when I control the device (open or close), it will report current status with cluster 0008 and attribute Id 0000

File: devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy
Comment (BYDengS) [reply devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy]: it's a good idear

Comment (dkirker) [reply devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy]: @BYDengS I believe what @tpmanley meant was to leave the code in `parse()` as is, but instead of having a function like `isWistar()` to have `usesLevelControl()`.

Comment (dkirker) [reply devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy]: @BYDengS So I guess really in this case, `isWistarDevice()` would become `usesLevelControl()` in the code before your last commit.

Comment (BYDengS) [reply devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy]: @dkirker ah,I misunderstood.

File: devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy
Comment (dkirker) [inline devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy]: I would flip this around so that you have:

```
if (usesWindowCovering()) {
	window covering stuff
} else {
	level stuff
}
```

File: devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy
Comment (dkirker) [inline devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy]: Let's just do
```
def usesWindowCovering() {
	device.getDataValue("manufacturer") != "Feibit Co.Ltd"
}
```

File: devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy
Comment (dkirker) [inline devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy]: @tpmanley Do you think we should maybe not check here for usesWindowCovering and just handle the messages that we receive from the device?

**Review by dkirker — COMMENTED** [review]
@BYDengS I left a couple of comments inline. Otherwise, looks pretty good!

File: devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy
Comment (tpmanley) [reply devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy]: @BYDengS This device has the following `inClusters`: `0000 0003 0004 0005 0006 0008 0102`. Please update the fingerprint

File: devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy
Comment (tpmanley) [reply devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy]: I think that's OK to handle everything as long as one device doesn't actually use both and they either provide conflicting information or both report at the exact same time and lead to the possibility of duplicate events

File: devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy:149]: So is it the case that the Wistar supports the Window Covering Cluster but not the GOTO_LIFT_PERCENTAGE command or the LIFT_PERCENTAGE attribute? If that's the case, perhaps it would be more clear to change `usesWindowCovering` to `supportsLiftPercentage`

File: devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy
Comment (BYDengS) [reply devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy]: Ok, I will update this

File: devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy
Comment (BYDengS) [reply devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy:149]: No, I have try the GOTO_LIFT_PERCENTAGE command and the LIFT_PERCENTAGE attribute, the device return useless information. I will update this.

**Comment by BYDengS** [comment]
@dkirker and @tpmanley , I have update the code according to comments, please help review again.

File: devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy:28]: The ticket makes it look like the join name should be `Wistar Curtain Motor CMJ`. If it's actually supposed to have the parenthesis then the ticket should be updated. There should at least be a space before the `(`.

File: devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy
Comment (tpmanley) [inline devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy]: I think this can now be replaced with `supportsLiftPercentage`

File: devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy
Comment (tpmanley) [reply devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy]: Does the Wistar report ATTRIBUTE_POSITION_LIFT but it's not correct or not report it at all? If not at all then I think you can remove `usesWindowCovering` and just handle every message that is received. If it reports the value and it's wrong so it needs to be ignored then I think you can replace `usesWindowCovering` with `supportsLiftPercentage` and then delete `usesWindowCovering`

File: devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy
Comment (BYDengS) [reply devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy:28]: https://smartthings.atlassian.net/browse/WWST-4447
in the catalog ticket, the join name should be same as `Setupapp Name: Wistar Curtain Motor(CMJ)  威仕达开合帘电机(CMJ)`

File: devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy
Comment (BYDengS) [reply devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy]: @tpmanley I have changed it, please review

File: devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy
Comment (tpmanley) [reply devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy]: @BYDengS ^

Comment (BYDengS) [reply devicetypes/smartthings/zigbee-window-shade.src/zigbee-window-shade.groovy]: @tpmanley  No, it not report it at all, I will delete it.

**Review by tpmanley — DISMISSED** [review]

**Comment by tpmanley** [comment]
@BYDengS changes look good to me now. The commits need to be squashed so you can either do a rebase and squash yourself and we can do a squash when we do the merge. This is currently against the staging branch which we usually do not do when the changes have the potential to affect existing devices like this change does. Is there QA lined up to verify the functionality of both the Wistar and at least one of the other two devices using this DTH before EOD Monday? If not then this PR should be redirected to `master` so that QA can have next week to test it.

**Comment by BYDengS** [comment]
@tpmanley,  @dkirker and @greens ,I have rebase it to one commit. Please review.

**Review by tpmanley — APPROVED** [review]

**Comment by tpmanley** [comment]
@BYDengS please let us know when this has passed QA. If this doesn't happen before the release tomorrow then we can merge it after the release.

**Comment by BYDengS** [comment]
@tpmanley  I have tested it with two read device (Dooya and Wistar) by self-publish this DTH,  it works fine. And next step it will be tested by our certification team in China STG environment. Please help merge it to China STG Server. Thank you.

**Review by dkirker — APPROVED** [review]
@BYDengS I'm checking on the pause() mappings to see what state those are in. Otherwise, this looks good.

**Comment by greens** [comment]
holding off on this until I hear from Donald

**Comment by dkirker** [comment]
@greens I believe @tpmanley told me last night he was going to wait for the deploy and then merge to staging. That will get it to staging for @BYDengS and allow some time in staging before going to production.

**Pull request merged**
